### ISSUE TITLE
functions created with `functools.partial` don't match `Callable`

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#6961]
**Describe the bug**
If passing a function using `functools.partial` to another function expecting a `Callable`, pyright reports an error

**Code or Screenshots**
```python
from functools import partial
from typing import Callable

basetwo = partial(int, base=2)


def foo(func: Callable[[object], object], value: str):
    print(func(value))


foo(basetwo, "1011")
```
reports
```text
\partialcallable.py:11:5 - error: Argument of type "partial[int]" cannot be assigned to parameter "func" of type "(object) -> object" in function "foo"
    Type "partial[int]" cannot be assigned to type "(object) -> object"
      Parameter 1: type "object" cannot be assigned to type "str | bytes | bytearray"
        Type "object" cannot be assigned to type "str | bytes | bytearray"
          "object" is incompatible with "str"
          "object" is incompatible with "bytes"
          "object" is incompatible with "bytearray" (reportGeneralTypeIssues)
```

Note that `mypy` accepts this.

**VS Code extension or command-line**

pyright 1.1.345 command line

**Comment by erictraut** [comment]
Pyright's behavior here is correct. This is a type violation, for the same reason that it would be a type violation to pass `int` as the first argument to `foo`.

```python
foo(int, 1)
```

Mypy doesn't generate an error in the original example because its handling of `functools.partial` is notably incomplete. Mypy does generate the expected error in my small snippet above when `partial` is not involved.

**Comment by erictraut** [comment]
You can fix the original example by changing it to the following:

```python
def foo(func: Callable[[str | bytes | bytearray], object], value: str):
    print(func(value))
```

**Comment by Dr-Irv** [comment]
So in the case of `pandas` where someone reported this, and the `Callable` takes any object as a parameter and returns any object as a result, what's the way to support a generic callable and a `partial` ?

**Comment by erictraut** [comment]
This is unrelated to `partial`. It's more fundamental than that. It's related to type compatibility rules for all callables in general. Input parameters for callables are always contravariant. The type annotation `func: Callable[[object], object]` refers to a callable that accepts a single input argument that is compatible with `object`. In other words, any compatible callback must be prepared to accept any `object` as an argument. However, the constructor for `int` does not accept `object`. It accepts only `str | bytes | bytearray`. 

I don't have enough context about the specifics of the problem you're trying to solve, but hopefully this explanation gets you unstuck.

**Comment by Dr-Irv** [comment]
OK, in our case, the API accepts a callable that takes in any kind of type, and returns any kind of type, so I think `Callable[[Any], Any]` will fix it.

Thanks for the explanation
